File: 0064-minimum-path-sum/0064-minimum-path-sum.py

class Solution:
    def minPathSum(self,grid: [List[List[int]]])->int:
        m = len(grid)
        n = len(grid[0])
        
        dp = [[float('inf')] * n for _ in range(m)]
        dp[0][0] = grid[0][0]
        
        def f(i,j):
            if i == 0 and j ==0:
                return dp[0][0]
            if i < 0 or j < 0:
                return float('inf')
            if dp[i][j] != float('inf'):
                return dp[i][j]
            up,left =float('inf'),float('inf')
            if i > 0:
                up = f(i-1,j)
            if j > 0:
                left = f(i,j-1)
                
            dp[i][j] = grid[i][j] + min(up,left)
            
            return dp[i][j]
            
        return f(m-1,n-1)

File: 0064-minimum-path-sum/0064-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int n = grid.length;
        int m = grid[0].length;
        
        // Space optimized to O(m) - only store current row
        int dp[] = new int[m];
        dp[0] = grid[0][0];
        
        // Initialize first row
        for(int j=1; j<m; j++){
            dp[j] = dp[j-1] + grid[0][j];
        }
        
        for(int i=1; i<n; i++){
            // Update first column for current row
            dp[0] = dp[0] + grid[i][0];
            
            for(int j=1; j<m; j++){
                // dp[j] (before update) = cell above
                // dp[j-1] = cell to the left
                dp[j] = grid[i][j] + Math.min(dp[j], dp[j-1]);
            }
        }
        return dp[m-1];
    }
}